sys/linux/bpf.txt

```diff
@@ -736,17 +729,15 @@ bpf_btf_strings {
 bpf_btf_strings_elem = 'a', '0', '_', '.', 0
 
 bpf_task_fd_query {
-# NEED: part of fields are input here and part are output. We can't express this yet (#245).
-	pid		pid
-	fd		fd_perf_base
-	flags		const[0, int32]
-	buf_len		bytesize[buf, int32]
-	buf		ptr64[in, string]
-# These are output fields:
-	prog_id		bpf_prog_id[opt]
-	fd_type		const[0, int32]
-	probe_offset	const[0, int64]
-	probe_addr	const[0, int64]
+	pid		pid	(in)
```

Comment:
Looking at this I am thinking what practice of using the direction we should establish.
There are basic primitives and capabilities, but I think we also need some practice and patterns of how to use the direction in typical situations.
We still have the direction on the pointer, in this case: `ptr[inout, bpf_task_fd_query]`. If we specify directions on all fields, what direction do we use on the pointer? Or do we use the pointer direction as "default" and don't specify direction for all fields?
Also I think instead of `const[0, int32]	(out)`, we should say `int32	(out)` because if it's `out` there is no point in adding more details wrt const/0.

Reply:
> If we specify directions on all fields, what direction do we use on the pointer? Or do we use the pointer direction as "default" and don't specify direction for all fields?
If we have a mix of directions in the struct, I just used `inout` at the ptr level but this isn't necessarily intiuitive. One approach might be (which requires some changes in the core code) to omit ptr direction for structs with all fields marked with directions. This would be easier on the eye and allow us to have additional checks at the compile time for such descriptions.

If we have a mix of directions and don't reuse a struct by changing the direction of the pointer, it seemed better to push all the directions to the per-field level for better readability. If the fields of different directions are close in count, which direction do we choose for the ptr, and what could be  good reason for this anyway?

Reply:
Thinking of this more I can't come up with any scheme right now. Let's proceed with your changes. I guess we need to use directions more to gather more experience.